pybamm/solvers/solution.py

```diff
@@ -471,16 +471,43 @@ def update(self, variables):
             variables = [variables]
         # Process
         for key in variables:
+            cumtrapz_ic = None
             pybamm.logger.debug("Post-processing {}".format(key))
             vars_pybamm = [model.variables_and_events[key] for model in self.all_models]
 
             # Iterate through all models, some may be in the list several times and
             # therefore only get set up once
             vars_casadi = []
-            for model, ys, inputs, var_pybamm in zip(
-                self.all_models, self.all_ys, self.all_inputs, vars_pybamm
+            for (i, (model, ys, inputs, var_pybamm)) in enumerate(
+                zip(self.all_models, self.all_ys, self.all_inputs, vars_pybamm)
             ):
-                if key in model._variables_casadi:
+                if isinstance(var_pybamm, pybamm.ExplicitTimeIntegral):
+                    cumtrapz_ic = var_pybamm.initial_condition
+                    if not pybamm.is_constant_and_can_evaluate(cumtrapz_ic):
+                        raise NotImplementedError(
+                            "Non-constant initial conditions has not been implemented."
+                        )
+                    else:
+                        cumtrapz_ic = cumtrapz_ic.evaluate()
+                    var_pybamm = var_pybamm.child
+                    t_MX = casadi.MX.sym("t")
+                    y_MX = casadi.MX.sym("y", ys.shape[0])
+                    inputs_MX_dict = {
+                        key: casadi.MX.sym("input", value.shape[0])
+                        for key, value in inputs.items()
+                    }
+                    inputs_MX = casadi.vertcat(*[p for p in inputs_MX_dict.values()])
+
+                    # Convert variable to casadi
+                    # Make all inputs symbolic first for converting to casadi
+                    var_sym = var_pybamm.to_casadi(t_MX, y_MX, inputs=inputs_MX_dict)
+
+                    var_casadi = casadi.Function(
+                        "variable", [t_MX, y_MX, inputs_MX], [var_sym]
+                    )
+                    model._variables_casadi[key] = var_casadi
+                    vars_pybamm[i] = var_pybamm
+                elif key in model._variables_casadi:
```

Comment:
Problem is that if that if statement is first, the thing returned will be incorrect, since the `cumtrapz` is located inside `ProcessedVariable`, the thing stored in `model._variables_casadi` is the integrand. Shouldn't `__getitem__` prevent this from being processed every time? 

But can fix duplication issue

Reply:
Oh yeah you're right, `__getitem__` will take care of it